### ISSUE TITLE
Testing of analysis and computational performance

### ISSUE THREAD

**Issue by SebastianZimmeck** [privacy-tech-lab/privacy-pioneer#282]
Once we are generally finished with our implementation, we need to test how good the analysis actually is. For example:

1. How many tracking pixel requests are classified as non-tracking pixel requests (false negatives)
2. How many non-tracking pixel requests are classified as tracking pixel requests (false positives)

So, we need a set of sites that have tracking pixels and a set that do not have those. I will come up with a test design. We need to do this testing for all our analysis functionalities (e.g., fine location, fingerprinting, ...). In the course of our testing we may need to adapt our techniques. If we cannot get one item to work sufficiently, we may also need to drop it. Once we are happy with the testing, we create a final test set with new sites/requests, and measure the performance (e.g., precision, recall, and F-1).

**Comment by SebastianZimmeck** [comment]
We may be able to start testing in about two weeks, at which point I should have a plan.

**Comment by SebastianZimmeck** [comment]
### 1. The Plan

We discussed today that it is time to move into testing. Here is what we need to do:
1. Come up with the testing methodology and set up any testing infrastructure if necessary
    - This issue
2. Fine-tuning and validation testing (i.e., test on a validation set of sites such that we are reasonably confident we get good results in the test) and do remaining fine-tuning, which may be informed by our validation tests
    - Issue #220 
    - Issue #336
    - Create validation set
3. Testing on the test set, i.e., on new unseen sites, to get the final performance numbers
    - Issue #289
    - Create test set

In principle, we need to be finished with all implementation that could affect the accuracy or computational performance analysis.

### 2. What do we test?

We test for (1) accuracy of detecting privacy practices and (2) for computational performance.

1. Accuracy
On the testing for accuracy, it is good to start with percentages, e.g., out of 100 sites in our test set 20 do fingerprinting, and for how many of those do we correctly detect the fingerprinting practice. However, to account for imbalanced datasets, e.g., it is a difference if we have 50 sites doing fingerprinting or 20, we want to measure [precision, recall, and F-1 score](https://en.wikipedia.org/wiki/Precision_and_recall). To calculate those, we need the following:
   - True positives: Site has actually fingerprinting and Privacy Pioneer finds fingerprinting
   - False positives: Site does not have actually fingerprinting and Privacy Pioneer finds fingerprinting 
   - True negatives: Site actually does not have fingerprinting and Privacy Pioneer does not find fingerprinting
   - False negatives: Site has not actually fingerprinting and Privacy Pioneer does find fingerprinting

2. Computational Performance
The second ---smaller item--- that we need to test for is computational performance. For that test we do not need precision, recall, etc. Rather, we would need to formalize a bit more the test regimen that @notowen333 already performed during his performance improvement journey. In other words, we would need something like: load 100 sites, measure how long it takes for each site to create the analysis results, say, measured in tenths of a second, maybe what the cpu usage is during that time, etc. This does not have to be sophisticated, but essentially have a sufficient data analysis to support the claim that Privacy Pioneer is usable for an average user browsing the web with an average machine.

Maybe, we can use the same sites from our test set browsing those and measure the time and load.

### 3. Why do we test?

The reason why we are testing computational performance is to make sure that an average user can actually use our extension.

The reason why we are testing accuracy is to make sure that our results are meaningful. We can display all kinds of labels, but if the user cannot trust that the labels indicate what actually happened, the analysis and the labels would be pointless. Testing on the sites in the test set allows us to extrapolate the accuracy of our measurements in the test set to unknown sites. Something like: With 95% confidence we can say that the accuracy results in our test will hold for any set of sites visited.

### 4. What do we test for in the accuracy testing?

We come up with our own definitions of what constitutes a certain privacy practice, and then we measure that. For example, we define that if an HTTP response body contains the string `fingerprint.js`, that string should qualify as evidence to label a site as "performing fingerprinting." If a site does not contain `fingerprint.js` or any other evidence that we define as indicative of fingerprinting, the site should be labeled as "not performing fingerprinting."

Obviously, our definitions should be corresponding to what Privacy Pioneer is actually doing. So, the above `fingerprint.js` example only makes sense if Privacy Pioneer is doing it that way. 

Also, our tests only claim in "mechanical" way certain code snippets were seen etc. Our tests do not relate to the question of whether `fingerprint.js` is actually a meaningful proxy for identifying fingerprinting. This latter question is one of definition and we address it at the definitional level. That is what the validation and tuning phase is for. In that phase we ask whether the correspondence of the evidence to what the labels are intended to express (e.g., `fingerprint.js` leads to `fingerprint` label makes sense). If we find sites that have for whatever reason `fingerprint.js` but in our opinion do not perform fingerprinting, we need to change our definition in that phase.

### 5. Establishing and Documenting Testing Methodology

This issue is about establishing and documenting the testing methodology (foremost the accuracy testing; depending how extensive the performance testing gets we need a new issue for that). In our paper we want to describe out testing methodology along the lines of: "We took a randomly selected set of 50 sites that we used as a validation set. We tuned our keyword ...  Then, we took 100 randomly selected new sites as a test set ... For each site tested we used the built-in browser developer tools to observe whether we found the string `fingerprint.js` in a site's response body ... We automated ... ."

So, the bottom line is to document how we test and what our setup is so that others can replicate it.

### 6. Two Important Points for the Testing Methodology

1. One very important point for testing is that the test set can only be run after the validation. What we see in the test set cannot inform our decisions for code changes. In order to be valid the test set must be fresh and unseen. A corollary of this is that if the results from our test set are not good enough ---and we should aim for 90%+ accuracy--- we have to create a new test set and iterate once more. This time we can use the now old test set as our validation set and use it to address the deficiencies. But we also need to create a new test set.

2. Another important point is that at the time we begin our testing we can no longer make (substantial) changes to our analysis code. Or, if we do such changes, we have to test again. So, for example, if add or change our analysis to include some ML techniques, we would need to test again. For changes that are not related to the analysis, though, we would not need to do so.

### 7. Keeping Track of Test Sets, Test Scripts, Test Protocols etc.

- Repo: I have set up a [new repo](https://github.com/privacy-tech-lab/privacy-pioneer-data-and-scripts) for data and scripts, in particular, as they relate to testing. For example, @notowen333 as you mentioned automation today, feel free to put any code for that there. 
  TODO: Please accept the invite and `watch` the repo.
- Google Drive: We can also use the usual Google Drive for documents such as Google Sheets etc.  Here is the link to a [Testing folder](https://drive.google.com/drive/folders/15bVoodSplBgcbGqiinEOJGh2UmbpdCbz?usp=sharing) I created. Feel free to put any test-related documents there. I have put one [Google Sheet in there as a starting point for the validation and test sets](https://docs.google.com/spreadsheets/d/1RFE8V1wyybySIIi5rC7JaRhC95mIe53e-WZmUPvu7sw/edit?usp=sharing). You can see where this is going ... 
   **TODO: @danielgoldelman, @notowen333, and @Lr-Brown, if you could put in there all your sites that you have used informally so far under the validation tab (e.g., sites that we know have tracking pixels, fingerprint.js, ...).** That will give us a start.

### 8. Details on Testing Methodology

In the coming weeks we will need to flesh out the details of our testing methodology, for example, how many sites to test on. The order is probably around 50 sites for the validation set (or whatever makes us comfortable to move on to the test set) and 100 for the test set. Other questions are: How do we find the sites? How do we test for the custom watchlist? ... 

As discussed today, @notowen333 will take the lead on testing.

If anything is unclear or there is something to say here, let's discuss ...

**Comment by SebastianZimmeck** [comment]
Per what I said [above](https://github.com/privacy-tech-lab/privacy-pioneer/issues/282#issuecomment-926223870), @notowen333, @danielgoldelman, and @Lr-Brown, if you could put in there all your sites that you have used informally so far under the [validation tab](https://docs.google.com/spreadsheets/d/1RFE8V1wyybySIIi5rC7JaRhC95mIe53e-WZmUPvu7sw/edit?usp=sharing) (e.g., sites that we know have tracking pixels, fingerprint.js, ...). That will give us a start. Also, please add a `1` in the `Site has` cells if the site has fingerprinting etc. (@danielgoldelman, I believe the current entries are already from you; if you could add the `Site has` information, that would be great.) 

I also want to emphasize that how we test is not set in stone. So, if you have other ideas or suggestions, let's discuss.

**Comment by SebastianZimmeck** [comment]
- **Where to get validation/test sites from?** -> The [Tranco list](https://tranco-list.eu/), which is a free and good list combining Alexa and other lists; maybe select from top to bottom or randomly. Possibly, we can also select sites from [FP-Inspector](https://uiowa-irl.github.io/FP-Inspector/) because those add external verification to our fingerprinting results.
- **To add external verification to our definitions** (and findings based on our definitions) we may be able to find resources that come to the same conclusion as PP. For example, if an external resource also comes to the conclusion that a site practices, say, fingerprinting, our findings receive additional weight. Fingerprinting seems to be the main practice where this may be helpful. Most of our results can be just validated by looking at them. For example, it will be pretty clear whether we have found a phone number or not (format, does the area code really exist, etc.). The Disconnect list is only somewhat an external verification because that is what we are using ourselves.

   External Justification
   - Monetization
      - Advertising: (Disconnect list)
      - Analytics: (Disconnect list)
      - Social Networking: (Disconnect list)
   - Location
      - Coarse Location: None
      - Fine Location: None, just check that the string includes GPS coordinates
      - ZIP Code: None, just check that the string includes a  ZIP code
      - Street Address: None, just check that the string includes a street address
      - City: None, just check that the the string includes a city
      - State: None, just check that the string includes a state
   - Tracking
      - Tracking Pixel: None
      - IP Address: None, just check that the string includes an IP address
      **- Browser Fingerprinting: (FingerprintingInvasive from Disconnect list), [FP-Inspector](https://uiowa-irl.github.io/FP-Inspector/)**
   - Watchlist
      - Phone Number: None, just check that the string includes a phone number
      - Email Address: None, just check that the string includes an email address
      - Custom Keywords: None, just check that the string includes a custom keyword

   @notowen333, @Lr-Brown, and @danielgoldelman please add to the list above other external verification sources if you have any. 

   I added an external verification column to fingerprinting in the validation/test set Google Sheet.

**Comment by SebastianZimmeck** [comment]
As discussed today, testing is on hold as we are digging deeper into the techniques for our various privacy practice analyses. We first have to settle on what privacy practices we analyze and which techniques we use. Especially, today we discussed NLP/ML techniques. I will open a new issue on that. Once we resolved the questions on what and how to analyze, we will revisit how we test.
